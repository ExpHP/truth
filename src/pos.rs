//! Source code locations (some parts borrowed from [qluon])
//!
//! [qluon]: https://github.com/gluon-lang/gluon/blob/master/base/src/pos.rs

use std::fmt;
use std::borrow::Cow;

pub type FileId = usize;
use codespan_reporting::{files as cs_files};
pub use codespan::Span;
pub use codespan::{ByteIndex as BytePos};

pub type Files = NonUtf8Files;

/// An implementation of [`codespan_reporting::files::Files`] adapted to non-UTF8 files.
#[derive(Debug, Clone)]
pub struct NonUtf8Files {
    inner: cs_files::SimpleFiles<String, String>,
}

impl NonUtf8Files {
    pub fn new() -> Self { NonUtf8Files { inner: cs_files::SimpleFiles::new() } }

    pub fn add(&mut self, name: impl AsRef<std::path::Path>, source: &[u8]) -> FileId {
        self.inner.add(
            name.as_ref().to_string_lossy().into_owned(),
            prepare_diagnostic_text_source(source).into(),
        )
    }
    pub fn source(&mut self, file_id: FileId) -> Result<&[u8], cs_files::Error> {
        <_ as cs_files::Files>::source(&self.inner, file_id)
            .map(|s| s.as_bytes())
    }
    pub fn get(&mut self, file_id: FileId) -> Result<&cs_files::SimpleFile<String, String>, cs_files::Error> {
        self.inner.get(file_id)
    }
}

impl<'a> cs_files::Files<'a> for NonUtf8Files {
    type FileId = FileId;
    type Name = String;
    type Source = &'a str;

    // Just delegate everything
    fn name(&self, file_id: FileId) -> Result<String, cs_files::Error> { self.inner.name(file_id) }

    fn source(&self, file_id: FileId) -> Result<&str, cs_files::Error> { self.inner.source(file_id) }

    fn line_index(&self, file_id: FileId, byte_index: usize) -> Result<usize, cs_files::Error> {
        self.inner.line_index(file_id, byte_index)
    }
    fn line_range(&self, file_id: FileId, line_index: usize) -> Result<std::ops::Range<usize>, cs_files::Error> {
        self.inner.line_range(file_id, line_index)
    }
}

/// A version of `from_utf8_lossy` that preserves byte positions.
///
/// The output of this is suitable for rendering spans in error messages.
///
/// It accomplishes this by using `?` as the replacement character, which only takes a single byte
/// and can thus easily fill arbitrarily-sized spaces, unlike `U+FFFD REPLACEMENT CHARACTER`
/// which takes three bytes.
fn prepare_diagnostic_text_source(s: &[u8]) -> Cow<str> {
    match std::str::from_utf8(s) {
        Ok(valid) => Cow::Borrowed(valid),
        Err(error) => {
            let mut remaining = s;
            let mut out = String::new();
            let mut res = Err(error);
            while let Err(error) = res {
                let (valid, after_valid) = remaining.split_at(error.valid_up_to());
                out.push_str(std::str::from_utf8(valid).expect("already validated"));

                let num_bad = error.error_len().unwrap_or(after_valid.len());
                for _ in 0..num_bad {
                    out.push('?');
                }
                remaining = &after_valid[num_bad..];
                res = std::str::from_utf8(remaining);
            }
            match res {
                Err(_) => unreachable!(),
                Ok(remaining_str) => out.push_str(remaining_str),
            }
            assert_eq!(s.len(), out.len());
            Cow::Owned(out)
        },
    }
}

#[test]
fn test_lossy_utf8() {
    let func = prepare_diagnostic_text_source;

    // valid UTF-8
    assert_eq!(func(b"ab\xF0\x9F\x92\x96cd"), "abðŸ’–cd");

    // invalid byte sequence...
    assert_eq!(func(b"\x80\xFFcd"), "??cd"); // ...at beginning
    assert_eq!(func(b"ab\x80\xFFcd"), "ab??cd"); // ...in middle
    assert_eq!(func(b"ab\x80\xFF"), "ab??"); // ...at end

    // incomplete character; byte 0b11110000 expects 3 more bytes after it.
    // (this is the case where Utf8Error::error_len() returns None)
    assert_eq!(func(b"ab\xF0\x80\x80"), "ab???");

    // unpaired surrogate
    // http://simonsapin.github.io/wtf-8/#surrogates-byte-sequences
    assert_eq!(func(b"ab\xED\xA3\xA4cd"), "ab???cd");

    // ambiguous case.  This begins with a 4-byte character starter byte, but returns to ascii after
    // 2 bytes. I'm not sure whether the documentation of `Utf8Error::error_len` is specified
    // well-enough to determine whether this would replace the two 'w' characters.
    let input = b"ab\xF0\x80wwcd";
    let output = func(input);
    assert_eq!(output.len(), input.len());
    assert_eq!(&output.as_bytes()[..2], &input[..2]);
    assert_eq!(&output.as_bytes()[2..2+2], b"??");
    assert_eq!(&output.as_bytes()[2+4..], &input[2+4..]);
}

/// An AST node with a span.  The span is not included in comparisons or hashes.
#[derive(Copy, Clone, Default)]
pub struct Spanned<T: ?Sized> {
    pub span: Span,
    pub value: T,
}

impl<T: fmt::Debug> fmt::Debug for Spanned<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Spanned")
            // format as a range instead of Span's derived Debug
            .field("span", &(self.span.start().0..self.span.end().0))
            .field("value", &self.value)
            .finish()
    }
}

impl<T> Spanned<T> {
    pub fn null_from<U: Into<T>>(value: U) -> Self {
        Spanned {
            span: Span::default(),
            value: value.into(),
        }
    }

    pub fn new_from<U: Into<T>>(span: Span, value: U) -> Self {
        Spanned { span, value: value.into() }
    }
}

impl<T> From<T> for Spanned<T> {
    fn from(value: T) -> Self {
        Spanned {
            span: Span::default(),
            value,
        }
    }
}

impl<T: ?Sized + Eq> Eq for Spanned<T> {}

impl<T: ?Sized + PartialEq> PartialEq for Spanned<T> {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}

impl<T: ?Sized + PartialEq> PartialEq<T> for Spanned<T> {
    fn eq(&self, other: &T) -> bool {
        self.value == *other
    }
}

impl<T: ?Sized + std::hash::Hash> std::hash::Hash for Spanned<T> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}

impl<T: ?Sized> std::ops::Deref for Spanned<T> {
    type Target = T;
    fn deref(&self) -> &T {
        &self.value
    }
}

impl<T: ?Sized> std::ops::DerefMut for Spanned<T> {
    fn deref_mut(&mut self) -> &mut T {
        &mut self.value
    }
}

impl<T: ?Sized, U: ?Sized> AsRef<U> for Spanned<T>
where
    T: AsRef<U>,
{
    fn as_ref(&self) -> &U {
        self.value.as_ref()
    }
}

impl<T> Spanned<T> {
    pub fn map<U>(self, mut f: impl FnMut(T) -> U) -> Spanned<U> {
        Spanned {
            span: self.span,
            value: f(self.value),
        }
    }
}

impl<T: ?Sized + fmt::Display> fmt::Display for Spanned<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", &self.value)
    }
}
