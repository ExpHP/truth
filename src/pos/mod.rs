//! Source code locations (some parts borrowed from [qluon])
//!
//! [qluon]: https://github.com/gluon-lang/gluon/blob/master/base/src/pos.rs

pub type FileId = Option<std::num::NonZeroU32>;
pub use codespan::{ByteIndex as BytePos, ByteOffset, RawIndex, RawOffset};

pub use span::{Sp, Span, HasSpan};
#[macro_use] mod span;

pub use source_map::Files;
mod source_map;

pub use spanned_str::SpannedStr;
mod spanned_str;
