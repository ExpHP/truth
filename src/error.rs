use crate::pos::{FileId, Span, HasSpan};

use codespan_reporting as cs;
type CsDiagnostic = cs::diagnostic::Diagnostic<FileId>;
type CsLabel = cs::diagnostic::Label<FileId>;

/// An error type that is intended to be pretty-printed through [`codespan_reporting`].
///
/// A `CompileError` may contain multiple errors.  It may even contain no errors!  This should
/// not happen when calling a function that returns a ``Result<T, CompileError>`, and is mainly
/// for use in code that attempts to gather errors from many sources (typically ending with a
/// call to [`CompileError::into_result`]).
#[derive(thiserror::Error, Debug)]
#[must_use = "A CompileError must be emitted or it will not be seen!"]
#[error("a diagnostic wasn't formatted. This is a bug! The diagnostic was: {:?}", .diagnostics)]
pub struct CompileError {
    diagnostics: Vec<Diagnostic>
}

impl CompileError {
    /// Zips two CompileError results, combining the errors if the both fail.
    pub fn join<A, B>(a: Result<A, CompileError>, b: Result<B, CompileError>) -> Result<(A, B), CompileError> {
        match (a, b) {
            (Ok(a), Ok(b)) => Ok((a, b)),
            (Err(e), Ok(_)) => Err(e),
            (Ok(_), Err(e)) => Err(e),
            (Err(mut a), Err(b)) => {
                a.append(b);
                Err(a)
            },
        }
    }
}

/// A single error in a [`CompileError`].  You can still add more labels to it.
///
/// It converts into [`CompileError`] using [`From`], so a `?` should suffice.
#[derive(Debug)]
pub struct Diagnostic {
    imp: CsDiagnostic,
}

impl Diagnostic {
    pub fn error() -> Self { Diagnostic { imp: CsDiagnostic::error() } }
    pub fn warning() -> Self { Diagnostic { imp: CsDiagnostic::warning() } }

    pub fn code(&mut self, code: String) -> &mut Self {
        self.imp.code = Some(code);
        self
    }

    pub fn message(&mut self, message: String) -> &mut Self {
        self.imp.message = message;
        self
    }

    /// Add a label of type 'primary'.
    pub fn primary(&mut self, span: impl HasSpan, message: String) -> &mut Self {
        let span = span.span();
        self.imp.labels.push(CsLabel::primary(span.file_id, span).with_message(message));
        self
    }

    /// Add a label of type 'secondary'.
    pub fn secondary(&mut self, span: impl HasSpan, message: String) -> &mut Self {
        let span = span.span();
        self.imp.labels.push(CsLabel::secondary(span.file_id, span).with_message(message));
        self
    }

    pub fn note(&mut self, message: String) -> &mut Self {
        self.imp.notes.push(message);
        self
    }
}

impl From<Diagnostic> for CompileError {
    fn from(d: Diagnostic) -> Self { CompileError { diagnostics: vec![d] } }
}

/// Error type used by parts of the codebase that don't have access to spans.
///
/// These parts of the codebase use `anyhow` to produce a single, fatal error message that may
/// include a chain of context.  This is always ultimately converted into a [`CompileError`]
/// shortly before being displayed to the user.
pub type SimpleError = anyhow::Error;

impl CompileError {
    pub fn new_empty() -> CompileError { CompileError { diagnostics: vec![] } }
    pub fn append(&mut self, mut other: CompileError) {
        self.diagnostics.append(&mut other.diagnostics);
    }
    /// Become an `Ok` if empty, and an `Err` otherwise.
    pub fn into_result<T>(self, value: T) -> Result<T, CompileError> {
        match self.diagnostics.len() {
            0 => Ok(value),
            _ => Err(self),
        }
    }
    pub fn into_result_with<T>(self, value: impl FnOnce() -> T) -> Result<T, CompileError> {
        match self.diagnostics.len() {
            0 => Ok(value()),
            _ => Err(self),
        }
    }
    pub fn error_count(&self) -> usize { self.diagnostics.len() }

    /// Drain all errors from this object and write them to the standard error stream.
    ///
    /// In order to render spans correctly, the [`crate::Files`] instance used to parse AST
    /// nodes is required.
    pub fn emit<'a>(&mut self, files: &'a impl cs::files::Files<'a, FileId=FileId>) -> Result<(), codespan_reporting::files::Error> {
        use cs::term::termcolor as tc;

        if std::env::var("_TRUTH_DEBUG__TEST").ok().as_deref() == Some("1") {
            let mut writer = tc::NoColor::new(vec![]);
            self.emit_to_writer(&mut writer, files)?;
            eprint!("{}", std::str::from_utf8(&writer.into_inner()).unwrap());
        } else {
            // typical
            let writer = tc::StandardStream::stderr(tc::ColorChoice::Auto);
            self.emit_to_writer(&mut writer.lock(), files)?;
        }
        Ok(())
    }

    fn emit_to_writer<'a>(
        &mut self,
        writer: &mut impl cs::term::termcolor::WriteColor,
        files: &'a impl cs::files::Files<'a, FileId=FileId>,
    ) -> Result<(), codespan_reporting::files::Error> {
        for e in self.diagnostics.drain(..) {
            cs::term::emit(writer, &*TERM_CONFIG, files, &e.imp).unwrap();
        }
        Ok(())
    }

    /// Render the errors to a text string with no color escapes.  Intended mainly for unit tests.
    pub fn to_string<'a>(&self, files: &'a impl cs::files::Files<'a, FileId=FileId>) -> Result<String, codespan_reporting::files::Error> {
        use codespan_reporting::term::{self, termcolor as tc};

        let mut writer = tc::NoColor::new(vec![]);
        for e in &self.diagnostics {
            term::emit(&mut writer, &*TERM_CONFIG, files, &e.imp).unwrap();
        }
        Ok(String::from_utf8(writer.into_inner()).expect("codespan crate wrote non-UTF8?!"))
    }

    /// Emit errors that contain no labels.
    ///
    /// It is a bug to call this when there is any possibility that the errors have labels.
    /// It should only be used when reading e.g. binary files, which have no position info.
    pub fn emit_nospans<'a>(&mut self) {
        // because there are no labels, we know the methods of the Files will never be used,
        // so we can use a dummy implementation.
        let result = self.emit(&crate::error::PanicFiles);

        // The only possible error here is an IO Error.  This would suggest STDERR is not writable,
        // which is hardly any reason to stop what we're doing, so just ignore all errors.
        drop(result);
    }
}

// -------------------------

lazy_static::lazy_static! {
    static ref TERM_CONFIG: codespan_reporting::term::Config = {
        let mut config = codespan_reporting::term::Config::default();
        // Make output closer to rustc. Fewer colors overall, looks better.
        config.styles.primary_label_error.set_intense(true);
        config.styles.secondary_label.set_intense(true);
        config.styles.line_number.set_intense(true);
        config.styles.source_border.set_intense(true);
        config
    };
}

#[macro_export]
macro_rules! _diagnostic {
    (
        @ $severity:ident,
        $(code=$code:literal,)? message($($message:tt)+)
        $(, primary( $primary_span:expr, $($primary_msg:tt)+ ) )*
        $(, secondary( $secondary_span:expr, $($secondary_msg:tt)+ ) )*
        $(, note( $($note_msg:tt)+ ) )*
        $(,)?
    ) => {{
        let mut d = $crate::error::Diagnostic::$severity();
        d.message(format!( $($message)+ ));
        $( d.code($code); )?
        $( d.primary(&$primary_span, format!( $($primary_msg)+ )); )*
        $( d.secondary(&$secondary_span, format!( $($secondary_msg)+ )); )*
        $( d.note(format!( $($note_msg)+ )); )*
        $crate::error::CompileError::from(d)
    }};
    ( // shorthand for message only
        @ $severity:ident,
        $message_fmt:literal $(, $message_arg:expr)* $(,)?
    ) => { $crate::_diagnostic!{
        @ $severity,
        message($message_fmt $(, $message_arg)*),
    }};
}

/// Generates a [`CompileError`] of severity `error`.
///
/// If you want to modify the error after the macro call (by dynamically adding labels/notes),
/// try the [`Diagnostic`] builder API instead.
#[macro_export]
macro_rules! error {
    ($($arg:tt)+) => { $crate::_diagnostic!(@error, $($arg)+) };
}

/// Generates a [`CompileError`] of severity `warning`.
///
/// If you want to modify the error after the macro call (by dynamically adding labels/notes),
/// try the [`Diagnostic`] builder API instead.
#[macro_export]
macro_rules! warning {
    ($($arg:tt)+) => { $crate::_diagnostic!(@warning, $($arg)+) };
}

/// Generates and immediately emits a `CompileError` of severity `warning` that has no labels.
#[macro_export]
macro_rules! fast_warning {
    ($($fmt_arg:tt)+) => {{
        $crate::warning!(message($($fmt_arg)+)).emit_nospans()
    }};
}

impl<'a> From<crate::parse::Error<'a>> for CompileError {
    fn from(e: crate::parse::Error<'a>) -> CompileError {
        use lalrpop_util::ParseError::*;

        match e {
            User { error } => error.into(),
            UnrecognizedEOF { location, ref expected } => error!(
                message("unexpected EOF"),
                primary(Span::from_locs(location, location), "unexpected EOF"),
                note("{}", DisplayExpected(expected)),
            ),
            UnrecognizedToken { token: (start, token, end), ref expected } => error!(
                message("unexpected token `{}`", token),
                primary(Span::from_locs(start, end), "unexpected token"),
                note("{}", DisplayExpected(expected)),
            ),
            ExtraToken { token: (start, token, end) } => error!(
                message("unexpected extra token `{}`", token),
                primary(Span::from_locs(start, end), "extra token"),
            ),
            InvalidToken { .. } => unreachable!("we don't use lalrpop's lexer"),
        }
    }
}

struct DisplayExpected<'a>(&'a [String]);
impl std::fmt::Display for DisplayExpected<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // copied from lalrpop_util
        if !self.0.is_empty() {
            writeln!(f)?;
            for (i, e) in self.0.iter().enumerate() {
                let sep = match i {
                    0 => "Expected one of",
                    _ if i < self.0.len() - 1 => ",",
                    // Last expected message to be written
                    _ => " or",
                };
                write!(f, "{} {}", sep, e)?;
            }
        }
        Ok(())
    }
}

impl From<anyhow::Error> for CompileError {
    fn from(e: anyhow::Error) -> CompileError {
        error!(message("{:#}", e)).into()
    }
}

impl From<std::io::Error> for CompileError {
    #[track_caller]
    fn from(e: std::io::Error) -> CompileError {
        // obviously not ideal but it's better than sitting there with a vague "file not found"
        // and no backtrace
        panic!("{}", e)
    }
}

impl From<crate::fmt::Error> for CompileError {
    fn from(e: crate::fmt::Error) -> CompileError {
        SimpleError::from(e).into()
    }
}

/// Trait for running an iterator and continuing after an `Err` to collect more errors.
pub trait GatherErrorIteratorExt {
    type OkItem;

    /// Collect an iterator, continuing after failure in order to gather more errors.
    ///
    /// If at least one of the items is `Err(_)`, it returns an `Err(_)` that concatenates all
    /// of the errors in the stream.  Otherwise, it returns `Ok(_)`.
    fn collect_with_recovery<B: std::iter::FromIterator<Self::OkItem>>(self) -> Result<B, CompileError>;
}

impl<Ts, T, E> GatherErrorIteratorExt for Ts
where
    Ts: Iterator<Item=Result<T, E>>,
    E: Into<CompileError>,
{
    type OkItem = T;

    fn collect_with_recovery<B: std::iter::FromIterator<T>>(self) -> Result<B, CompileError> {
        let mut errors = CompileError::new_empty();
        let out = self.filter_map(|r| match r {
            Ok(x) => Some(x),
            Err(e) => { errors.append(e.into()); None }
        }).collect();

        errors.into_result(out)
    }
}

#[test]
fn test_collect_with_recovery() {
    // straightforward usage
    let result = (0..10).map(|x| match x % 2 {
        0 => Ok(x),
        1 => Err(error!(message("odd number: {}", x))),
        _ => unreachable!(),
    }).collect_with_recovery::<Vec<_>>();
    assert_eq!(result.unwrap_err().error_count(), 5);

    // collecting into () for side-effects
    let mut vec = vec![];
    let result = (0..10).map(|x| match x % 2 {
        0 => {
            vec.push(x);
            Ok(())
        },
        1 => Err(error!(message("odd number: {}", x))),
        _ => unreachable!(),
    }).collect_with_recovery::<()>();
    assert_eq!(vec, vec![0, 2, 4, 6, 8]);
    assert_eq!(result.unwrap_err().error_count(), 5);
}

// -------------------------

/// Implementation of [`codespan_reporting::files::Files`] where all methods panic.
#[doc(hidden)]
pub struct PanicFiles;

#[doc(hidden)]
pub type CsResult<T> = Result<T, codespan_reporting::files::Error>;

impl<'a> codespan_reporting::files::Files<'a> for PanicFiles {
    type FileId = FileId;
    type Name = &'static str;
    type Source = &'static str;

    fn name(&'a self, _: FileId) -> CsResult<Self::Name> { unreachable!("span in emit_nospans!") }
    fn source(&'a self, _: FileId) -> CsResult<Self::Source> { unreachable!("span in emit_nospans!") }
    fn line_index(&'a self, _: FileId, _: usize) -> CsResult<usize> { unreachable!("span in emit_nospans!") }
    fn line_range(&'a self, _: FileId, _: usize) -> CsResult<std::ops::Range<usize>> { unreachable!("span in emit_nospans!") }
}

// -------------------------

/// Utility that makes it easier to apply `anyhow::Context` to all errors in a region of code.
///
/// Basically, the most straightforward way to have many error paths apply the same context is to
/// put them in an IIFE.  However, (a) IIFEs can be confusing to read, and (b) you'd typically be
/// forced to explicitly annotate the return type due to the nested usage of `?`.
///
/// Basically,
/// ```text
/// group_anyhow(|| {
///     ...
/// }).with_context(|| { ... })
/// ```
/// is nicer than
/// ```text
/// (|| -> Result<_, anyhow::Error> {
///     ...
/// })().with_context(|| { ... })
/// ```
pub(crate) fn group_anyhow<T>(
    func: impl FnOnce() -> Result<T, anyhow::Error>,
) -> Result<T, anyhow::Error> {
    func()
}
