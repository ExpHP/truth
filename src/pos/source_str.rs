use regex::Regex;

use crate::pos::{FileId, Sp, Span};

/// Represents a piece of input source text with a known span.
///
/// Provides operations for slicing this text in a manner that also slices the span.
///
/// The text stored in [`SourceStr`] should be exactly the input source text, so that the length of
/// the [`Span`] is identical to the length of the string. Contrast this with [`Sp`]`<String>`, whose
/// span length and string length may differ due to e.g. escape sequences and inclusion/exclusion of
/// delimiters like `"`.
#[derive(Debug, Clone, Copy)]
pub struct SourceStr<'a> {
    pub str: &'a str,
    file_id: FileId,
    start_pos: usize,
}

lazy_static::lazy_static! {
    static ref LINE_END_RE: Regex = Regex::new(r#"\r\n?|\n"#).unwrap();
}

impl<'a> SourceStr<'a> {
    /// Constructs a [`SourceStr`] from a string that isn't in the source map.
    ///
    /// This is primarily meant to facilitate writing tests.  The spans from this will
    /// have a [`FileId`] of `None`, meaning that diagnostics may panic.
    pub(crate) fn new_null(str: &'a str) -> Self {
        Self::from_full_source(None, str)
    }

    /// Construct from the full source text associated with a [`FileId`].
    pub fn from_full_source(file_id: FileId, str: &'a str) -> Self {
        SourceStr { str, file_id, start_pos: 0 }
    }

    /// Construct from a source string and its corresponding span.
    pub fn from_span(span: Span, str: &'a str) -> Self {
        println!("{:?}", span);
        println!("{:?}", str);
        assert_eq!(span.len(), str.len(), "input str is not original source text, or span is wrong!");
        SourceStr {
            str,
            file_id: span.file_id,
            start_pos: span.start.into(),
        }
    }

    pub fn span(&self) -> Span {
        Span {
            file_id: self.file_id,
            start: (self.start_pos as u32).into(),
            end: ((self.start_pos + self.str.len()) as u32).into(),
        }
    }
}

/// Helper functions for some handwritten parsers.
impl<'a> SourceStr<'a> {
    /// Return a line (with the line terminator) and remove it from this string.
    pub fn split_next_line(&mut self) -> Option<SourceStr<'a>> {
        LINE_END_RE.find(self.str)
            .map(|m| self.split_off_front(m.end()))
            .or_else(|| {
                // last line may be missing terminator
                match self.str.len() {
                    0 => None,
                    n => Some(self.split_off_front(n)),
                }
            })
    }

    /// Return the first character and remove it from this string.
    pub fn split_next_char(&mut self) -> Option<Sp<char>> {
        self.str.chars().next().map(|char| {
            let char_str = self.split_off_front(char.len_utf8());
            sp!(char_str.span() => char)
        })
    }

    /// Return a copy with the first `index` characters and remove them from `self`.
    fn split_off_front(&mut self, index: usize) -> SourceStr<'a> {
        let out = self.slice_to(index);
        *self = self.slice_from(index);
        out
    }

    pub fn trim(&self) -> SourceStr<'a> {
        self.trim_end().trim_start()
    }

    pub fn trim_end(&self) -> SourceStr<'a> {
        self.slice_to(self.str.trim_end().len())
    }

    pub fn trim_start(&self) -> SourceStr<'a> {
        let ws_len = self.len() - self.str.trim_start().len();
        self.slice_from(ws_len)
    }

    pub fn slice(&self, index: std::ops::Range<usize>) -> SourceStr<'a> {
        self.slice_to(index.end).slice_from(index.start)
    }

    pub fn slice_to(&self, index: usize) -> SourceStr<'a> {
        SourceStr { str: &self.str[..index], ..*self }
    }

    pub fn slice_from(&self, index: usize) -> SourceStr<'a> {
        SourceStr {
            str: &self.str[index..],
            start_pos: self.start_pos + index,
            file_id: self.file_id,
        }
    }
}

impl std::ops::Deref for SourceStr<'_> {
    type Target = str;

    fn deref(&self) -> &str { self.str }
}

#[test]
fn test_trim() {
    // nontrivial
    let input = SourceStr { file_id: None, str: "  aaf io    ", start_pos: 17 };
    let trimmed = input.trim();
    assert_eq!(trimmed.start_pos, 19);
    assert_eq!(trimmed.str, "aaf io");

    // trivial
    let input = SourceStr { file_id: None, str: "aaf io", start_pos: 17 };
    let trimmed = input.trim();
    assert_eq!(trimmed.start_pos, 17);
    assert_eq!(trimmed.str, "aaf io");

    // all whitespace; edge case where a bad implementation could double-count
    let input = SourceStr { file_id: None, str: "   ", start_pos: 17 };
    let trimmed = input.trim();
    assert!(17 <= trimmed.start_pos && trimmed.start_pos <= 17 + input.len());
    assert_eq!(trimmed.str, "");
}
