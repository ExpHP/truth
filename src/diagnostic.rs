use codespan_reporting as cs;

use crate::pos::Files;
use crate::error::CompileError;

pub struct DiagnosticEmitter<'a> {
    files: &'a Files,
    config: &'a cs::term::Config,
    writer: &'a mut (dyn cs::term::termcolor::WriteColor + 'a),
}

impl<'a> DiagnosticEmitter<'a> {
    pub fn new(config: &'a cs::term::Config, files: &'a Files, writer: &'a mut impl cs::term::termcolor::WriteColor) -> Self {
        DiagnosticEmitter { config, files, writer }
    }

    /// Emit a diagnostic.
    pub fn emit(&mut self, diag: CompileError) {
        diag.emit_to_writer(self.writer, self.files, self.config);
        ErrorOccurred
    }

    /// Emit a diagnostic and return an `Err` with an empty [`CompileError`].
    ///
    /// Use this as `return emitter.fail_with(d);` or `emitter.fail_with(d)?`.
    /// Its purpose is to help reduce defects where one forgets to return an `Err` variant
    /// after emitting a fatal error.  (for this to happen, one would now have to
    /// specifically call the less popular [`Self::emit`] method instead)
    pub fn fail_with<T>(&mut self, diag: CompileError) -> Result<T, CompileError> {
        self.emit(diag);
        Err(CompileError::new())
    }
}
