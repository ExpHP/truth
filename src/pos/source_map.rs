use std::borrow::Cow;
use std::num::NonZeroU32;
use std::cell::RefCell;
use std::rc::Rc;

use codespan_reporting::{files as cs_files};

use crate::pos::{FileId, Span, BytePos};
use crate::diagnostic::Diagnostic;
use crate::debug_info;

/// An implementation of [`codespan_reporting::files::Files`] for `truth`.
///
/// This is the type responsible for keeping track of source text so that snippets can be displayed
/// in diagnostic error messages.
#[derive(Debug, Clone)]
pub struct Files {
    inner: RefCell<FilesImpl>,
}

#[derive(Debug, Clone)]
pub struct FilesImpl {
    files: cs_files::SimpleFiles<String, Rc<str>>,
    max_file_id: FileId,
}

impl Files {
    pub fn new() -> Self { Files { inner: RefCell::new(FilesImpl {
        files: cs_files::SimpleFiles::new(),
        max_file_id: None,
    }) } }

    /// Add a piece of source text to the database, and give it a name (usually a filepath)
    /// which will appear in error messages.  Also validate the source as UTF-8.
    ///
    /// The name does not need to be a valid path or even unique; for instance, it is common to use
    /// the name `"<input>"` for source text not associated with any file.
    pub fn add(&self, name: &str, source: &[u8]) -> Result<(FileId, Rc<str>), Diagnostic> {
        let utf8_cow = prepare_diagnostic_text_source(source);
        let rc_source: Rc<str> = utf8_cow[..].into();

        let mut inner = self.inner.borrow_mut();
        let file_id = Self::shift_file_id(inner.files.add(name.to_owned(), rc_source.clone()));
        inner.max_file_id = file_id;

        // the cow is borrowed iff the input was valid UTF-8
        if let Cow::Owned(_) = utf8_cow {
            let err = std::str::from_utf8(source).unwrap_err();
            let pos = err.valid_up_to();
            return Err(error!(
                message("invalid UTF-8"),
                primary(Span::new(file_id, BytePos(pos as _), BytePos(pos as _)), "not valid UTF-8"),
                note("truth expects all input script files to be UTF-8 regardless of the output encoding"),
            ));
        }

        Ok((file_id, rc_source))
    }

    fn unshift_file_id(file_id: FileId) -> Result<usize, cs_files::Error> {
        // produce Error on file_id = None; such spans aren't fit for diagnostics
        let file_id: u32 = file_id.ok_or(cs_files::Error::FileMissing)?.into();
        Ok(file_id as usize - 1)
    }

    fn shift_file_id(file_id: usize) -> FileId {
        NonZeroU32::new(file_id as u32 + 1)
    }

    fn file_ids(&self) -> impl Iterator<Item=FileId> {
        let max_file_id = self.inner.borrow().max_file_id.map_or(0, |x| x.get());
        (1..=max_file_id).map(|x| Some(NonZeroU32::new(x).unwrap()))
    }

    pub fn debug_info(&self) -> Vec<debug_info::SourceFile> {
        self.file_ids().map(|file_id| {
            debug_info::SourceFile {
                id: file_id.map_or(0, |x| x.get() as i32),
                name: cs_files::Files::name(self, file_id).unwrap(),
            }
        }).collect()
    }
}

/// This implementation provides source text that has been lossily modified to be valid UTF-8,
/// and which should only be used for diagnostic purposes.
impl<'a> cs_files::Files<'a> for Files {
    type FileId = FileId;
    type Name = String;
    type Source = Rc<str>;

    // Just delegate everything
    fn name(&self, file_id: FileId) -> Result<String, cs_files::Error> {
        self.inner.borrow().files.name(Self::unshift_file_id(file_id)?)
    }

    fn source(&self, file_id: FileId) -> Result<Rc<str>, cs_files::Error> {
        Ok(self.inner.borrow().files.get(Self::unshift_file_id(file_id)?)?.source().clone())
    }

    fn line_index(&self, file_id: FileId, byte_index: usize) -> Result<usize, cs_files::Error> {
        self.inner.borrow().files.line_index(Self::unshift_file_id(file_id)?, byte_index)
    }
    fn line_range(&self, file_id: FileId, line_index: usize) -> Result<std::ops::Range<usize>, cs_files::Error> {
        self.inner.borrow().files.line_range(Self::unshift_file_id(file_id)?, line_index)
    }
}

/// Obtain a UTF-8 version of the source that is suitable for rendering spans in error messages
/// for potentially non-UTF8 text.
fn prepare_diagnostic_text_source(s: &[u8]) -> Cow<'_, str> {
    // Back when truth allowed scripts to be Shift-JIS, we had to worry about the replacement character
    // messing up byte offsets, and so this was more complicated.
    //
    // Now that we require UTF-8, the only possible error that needs to be rendered from a non-UTF-8 file
    // is an error at the FIRST appearance of non-UTF8 data; thus the byte offsets will be just fine.
    String::from_utf8_lossy(s)
}
