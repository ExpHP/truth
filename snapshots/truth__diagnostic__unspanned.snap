---
source: src/diagnostic.rs
expression: stderr
---
warning: a.txt: thing 3: while eating a sub: blah 20


